workers/scanner.go

```diff
@@ -0,0 +1,151 @@
+	- unlock it
+
+ PHASE 2:
+ - calculate server load by getting the total number and size of files pinned by each server
+ - only pin underpinned files if the current server is in the lowest 20% of servers, otherwise exit before scanning further
+
+ PHASE 3:
+ - add a second scanner which looks for skylinks which should be unpinned and unpins them from the local skyd.
+*/
+
+var (
+	// SleepBetweenPins defines how long we'll sleep between pinning files.
+	// We want to add this sleep in order to prevent a single server from
+	// grabbing all underpinned files and overloading itself. We also want to
+	// allow for some time for the newly pinned files to reach full redundancy
+	// before we pin more files.
+	SleepBetweenPins = build.Select(
+		build.Var{
+			Standard: 10 * time.Second,
+			Dev:      time.Second,
+			Testing:  time.Millisecond,
+		}).(time.Duration)
+
+	// SleepBetweenScans defines how often we'll scan the DB for underpinned
+	// skylinks.
+	SleepBetweenScans = build.Select(build.Var{
+		// In production we want to use a prime number of hours, so we can
+		// de-sync the scan and the sweeps.
+		Standard: 19 * time.Hour,
+		Dev:      10 * time.Second,
+		Testing:  100 * time.Millisecond,
+	}).(time.Duration)
+)
+
+type (
+	// Scanner is a background worker that periodically scans the database for
+	// underpinned skylinks. Once an underpinned skylink is found (and it's not
+	// being pinned by the local server already), Scanner pins it to the local
+	// skyd.
+	Scanner struct {
+		staticDB            *database.DB
+		staticLogger        *logrus.Logger
+		staticMinNumPinners int
+		staticServerName    string
+		staticSkydClient    skyd.ClientInterface
+		staticTG            *threadgroup.ThreadGroup
+	}
+)
+
+// NewScanner creates a new Scanner instance.
+func NewScanner(db *database.DB, logger *logrus.Logger, minNumPinners int, serverName string, skydClient skyd.ClientInterface, tg *threadgroup.ThreadGroup) *Scanner {
+	return &Scanner{
+		staticDB:            db,
+		staticLogger:        logger,
+		staticMinNumPinners: minNumPinners,
+		staticServerName:    serverName,
+		staticSkydClient:    skydClient,
+		staticTG:            tg,
+	}
+}
+
+// Start launches the background worker thread that scans the DB for underpinned
+// skylinks.
+func (s *Scanner) Start() error {
+	err := s.staticTG.Add()
+	if err != nil {
+		return err
+	}
+
+	go s.threadedScanAndPin()
+
+	return nil
+}
+
+// threadedScanAndPin defines the scanning operation of Scanner.
+func (s *Scanner) threadedScanAndPin() {
+	defer s.staticTG.Done()
+
+	// Main execution loop, goes on forever while the service is running.
+	for {
+		s.pinUnderpinnedSkylinks()
+		// Sleep between database scans.
+		select {
+		case <-time.After(SleepBetweenScans):
+		case <-s.staticTG.StopChan():
+			return
+		}
+	}
+}
+
+// pinUnderpinnedSkylinks loops over all underpinned skylinks and pin them.
+func (s *Scanner) pinUnderpinnedSkylinks() {
+	for s.findAndPinOneUnderpinnedSkylink() {
```

Comment:
I think the plan is to only grab one underpinned skylink and pin it. Then in the meantime, another server might pick up the next link or a new link might become underpinned. So we just do it one-by-one instead of fetching and locking all of them and then iterating through them.

Reply:
What Chris said is correct. I only want to clarify that we don't need the sleep after pinning, we want it there in order to give skyd a little time to upload the file to its full redundancy before pinning another one. We are in no hurry to finish this process as soon as possible, we'd rather take a little longer than causing unnecessary load on the server, which will happen if we rapidly pin a number of large skylinks which need to be uploaded to their full redundancy level.

Reply:
I think you both misread my code example. All servers can fetch the `underpinned` list of documents just fine without holding a lock on them. Then you just iterate over them and acquire a lock on one individual skylink. When you have that lock, the first thing you do is check whether it's still underpinned (another server might have gotten to it already) if so just continue.

Reply:
not what I meant but doesn't matter, already approved this PR I don't mind the current approach

Reply:
Well, this is not great. I don't feel right just dismissing your feedback because I didn't understand it. Maybe we can take it up on Discord later?